packages/web-frontend/cypress/scripts/generateConfig/configSchema.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { yup, yupUtils } from '@tupaia/utils';
+
+import { SNAPSHOT_TYPES } from '../../constants';
+
+const { oneOfType, oneOrArrayOf } = yupUtils;
+
+const shape = shapeConfig => yup.object(shapeConfig).noUnknown();
+
+const arrayOf = values => yup.array().of(yup.mixed().oneOf(values));
```

Comment:
I actually wrapped them to make the validation more readable! Without it, the schema becomes harder to parse, in my opinion...

Will post an example of inline statements, then we can decide together!

Reply:
About using `mixed`: this is just because the `oneOf` method is a child of `mixed`, not `yup` - see https://github.com/jquense/yup#mixedoneofarrayofvalues-arrayany-message-string--function-schema-alias-equals

Reply:
```js
/**
 * Tupaia
 * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
 */

import { yup, yupUtils } from '@tupaia/utils';

import { SNAPSHOT_TYPES } from '../../constants';

const { oneOfType, oneOrArrayOf } = yupUtils;

export const configSchema = yup.object({
  dashboardReports: yup
    .object({
      allowEmptyResponse: yup.boolean(),
      filter: yup
        .object({
          code: oneOrArrayOf(yup.string()),
          project: oneOrArrayOf(yup.string()),
          orgUnit: oneOrArrayOf(yup.string()),
          dashboardGroup: oneOrArrayOf(yup.string()),
          startDate: oneOrArrayOf(yup.string()),
          endDate: oneOrArrayOf(yup.string()),
          dataBuilder: oneOrArrayOf(yup.string()),
        })
        .noUnknown(),
      snapshotTypes: yup
        .array()
        .of(yup.mixed().oneOf([SNAPSHOT_TYPES.RESPONSE_DATA, SNAPSHOT_TYPES.HTML]))
        .min(1)
        .required(),
      urlFiles: yup.array().of(yup.string()),
      urls: yup.array().of(oneOfType([yup.string(), yup.object()])),
    })
    .noUnknown(),
  mapOverlays: yup.object({
    allowEmptyResponse: yup.boolean(),
    filter: yup
      .object({
        id: oneOrArrayOf(yup.string()),
        project: oneOrArrayOf(yup.string()),
        orgUnit: oneOrArrayOf(yup.string()),
        measureBuilder: oneOrArrayOf(yup.string()),
      })
      .noUnknown(),
    snapshotTypes: yup
      .array()
      .of(yup.mixed().oneOf([SNAPSHOT_TYPES.RESPONSE_DATA]))
      .min(1)
      .required(),
    urlFiles: yup.array().of(yup.string()),
    urlGenerationOptions: yup
      .object({
        id: oneOrArrayOf(yup.string()),
        project: oneOrArrayOf(yup.string()),
        orgUnit: oneOrArrayOf(yup.string()),
      })
      .noUnknown(),
    urls: yup.array().of(oneOfType([yup.string(), yup.object()])),
  }),
});
```

Reply:
Perhaps `strings`, `stringOrStrings` and `stringsOrObjects` are not worth it, but I feel that `arrayOf` is - we could add the latter in `@tupaia/utils/src/validation/yup.js` - thoughts @edmofro ?

Reply:
Good thoughts @kael89, and that compromise of wrapping up to a point sounds good. 

I would just like to get @mcccclean's thoughts on it though because this could set a precedent. 

@mcccclean in Tamanu what is your approach? Do you wrap up `yup` type chains into functions?

Pros:
- More readable and concise
Cons:
- Need to understand one extra thing when reading inline
- Potentially less consistent with Tamanu (and also google)?

Reply:
We actually had this same discussion in Tamanu recently! (although re a much smaller change, so I didn't give it as much thought). The outcome of that was, basically: yes, custom yup validators _should_ be wrapped in functions. The primary thought behind it was just that `yup.string()` and `myCustomYupValidator()` really should have a consistent signature, even if the custom validator doesn't accept any options.

(I guess it's also the case that if it _does_ end up taking options in the future, it doesn't require a refactor!)

Reply:
Thanks for the input @mcccclean, good to know what Tamanu's doing! Let's stick with that strategy @kael89 (which I think you have here :-))